File: Fortran/3_ex/fibonacci_ite.f90
program fibonacci_ite
    implicit none
    
    INTEGER :: A,B,S1,S2,i
    DOUBLE PRECISION :: start1, finish1, start2, finish2
    INTEGER :: C(0:1) !Defining an array ----> do not forget


    print*, 'Fibonacci s indice value for n = 50'
    write(*,'(/A)')

    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    !Variable iterative fibonacci 
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!

    A = 0 
    B = 1
    S1 = 0

    call cpu_time(start1)
    do i = 1,45

        A = S1
        S1 = A + B
        B = A

    end do 
    call cpu_time(finish1)

    print*,'Variable iterative fibonacci'
    print*, 'The number is: ',S1
    print*, finish1 - start1

    write(*,'(/A)')

    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    !Array iterative fibonacci 
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!

    C(0) = 0
    C(1) = 1

    S2 = 0
    
    !print*, C !--> Just to be more clear

    call cpu_time(start2)
    do i = 1,45

        C(0) = S2
        S2 = sum(C)
        C(1) = C(0)

    end do
    call cpu_time(finish2)

    print*,'Array iterative fibonacci'
    print*, 'The number is: ',S2
    print *, finish2 - start2
    

end program fibonacci_ite

File: Julia/3_ex/fibonacci_ite.jl
# Undergraduate Student: Arturo Burgos
# Professor: João Rodrigo Andrade
# Federal University of Uberlândia - UFU, Fluid Mechanics Laboratory - MFLab, Block 5P, Uberlândia, MG, Brazil


# Third exercise: Fibonacci sequence - by a common loop

using Printf
using Statistics
using CPUTime

println("Choose a indice value for n: ")
N = chomp(readline())
n = parse(Int32,N)
println("\n")

Fbn = 0
A = 0
B = 1

##############################
# Variable iterative fibonacci  --> Worse performance, why?
##############################

function varloop(F1,F2,S,k)
    for i = 1:k
        F1 = S
        S = F1 + F2
        F2 = F1
    end
    return S
    
end

Fbn = varloop(A,B,Fbn,n)

println("The correspond indices number is: ",Fbn)



@time @CPUtime varloop(A,B,Fbn,n)

println("\n")

###########################
# Array iterative fibonacci  --> Better performance, why?
###########################

Fbn = 0
C = [0 1]

function arrayloop(A,S,k)
    
    for i = 1:k
        A[1]=S
        S = sum(A)
        A[2] = A[1]
    end
    return S

end

Fbn = arrayloop(C,Fbn,n)

println("The correspond indices number is: ",Fbn)

@time @CPUtime arrayloop(C,Fbn,n)

